### DISCUSSION TITLE
Set default headers for calling downstream API's

### DISCUSSION THREAD

**Discussion by JaronGG** [AzureAD/microsoft-identity-web#1370] [Q&A]
I am working on a .NET Core 5 Web API that calls downstream Web API's. Calling these Web API's will be done through an API Manager. This manager requires that a subscription key to be present in the request.  I am using the `IDownstreamWebApi` configured in the startup to make HTTP calls like this.

```
            var response = await _downstreamWebApi.CallWebApiForAppAsync("downstreamAPI", options => {
                options.HttpMethod = HttpMethod.Get;      
                options.RelativePath = $"api/controller/{request.Id}";            
            });
```

I would like to have the subscription key configured or added here in the call. So my question is, is there a way to configure a default request header for the downstream web API or add these to the request? I can't seem to find anything about default header or extra query parameters.

Thank in advance!

**Comment by BederBourahmah** [comment]
I would like to know the same.

In my case, I am calling MS Graph using `IDownstreamWebApi` because I couldn't find a way for the the graph client to be introduced through dependency injection that would allow me to authenticate as an application and not a user. That's a whole different issue though.

I need to set the `ConsistencyLevel` header to `eventual` to use the search functionality, and it doesn't look like I can do that with `CallWebApiForAppAsync` . I will be falling back to calling it without the helping class to get this implemented, but I am pretty disappointed that we don't have the ability to control headers with `CallWebApiForAppAsync`.

**Reply by jmprieur** [reply]
> I would like to know the same.
> 
> In my case, I am calling MS Graph using `IDownstreamWebApi` because I couldn't find a way for the the graph client to be introduced through dependency injection that would allow me to authenticate as an application and not a user. That's a whole different issue though.
> 
Did you try to use the modifiers on the graph request? `.WithAppOnly()`

https://github.com/AzureAD/microsoft-identity-web/blob/b933e8a26f87c9f6fd4fa077ac7ca94026764655/tests/WebAppCallsMicrosoftGraph/Pages/AllApps.cshtml.cs#L32

**Reply by BederBourahmah** [reply]
This looks like exactly what I need. I'll try it first thing tomorrow. Was this available in MSDN docs or the repo docs and I missed it?

**Comment by jmprieur** [comment]
Modifying the headers is addressed n https://github.com/AzureAD/microsoft-identity-web/issues/1063 which is done.
It will be in Microsoft.Identity.Web 1.22.0, which should be released tomorrow.
cc: @jennyf19

**Reply by BederBourahmah** [reply]
You guys are awesome!
